### PR TITLE
Add OutputDebugString() to debug() strings on Windows

### PR DESCRIPTION
**Pull request by erik-smit**
Currently `debug()` strings go to stdout/stderr on Windows, which means they're usually not visible.

### CODE REVIEW

File: gum/gum.c [hunk 1]
```diff
@@ -16,6 +16,9 @@
 #include <ffi.h>
 #include <glib-object.h>
 #include <gio/gio.h>
+#ifdef HAVE_WINDOWS
+  #include <windows.h>
```
Comment (oleavr) [inline gum/gum.c]: We currently indent includes like this:
```C
#ifdef HAVE_WINDOWS
# include <windows.h>
#endif
```

File: gum/gum.c [hunk 2]
```diff
@@ -506,6 +509,18 @@ gum_on_log_message (const gchar * log_domain,
 
     return;
   }
+#else
```
Comment (oleavr) [inline gum/gum.c]: I'd say the stdio fallback doesn't make any sense on Windows – it only does on i/macOS where some processes may not have `CoreFoundation` loaded. So we should move this to the top and do like:

```C
#if defined (HAVE_WINDOWS)
...
#elif defined (HAVE_ANDROID)
...
#else
# ifdef HAVE_DARWIN
```

File: gum/gum.c [hunk 3]
```diff
@@ -506,6 +509,18 @@ gum_on_log_message (const gchar * log_domain,
...
+# ifdef HAVE_WINDOWS
+  gunichar2* utf16_message;
```
Comment (oleavr) [inline gum/gum.c]: Missing space before `*`.

File: gum/gum.c [hunk 4]
```diff
@@ -506,6 +509,18 @@ gum_on_log_message (const gchar * log_domain,
...
+  utf16_message = g_utf8_to_utf16 (message, -1, NULL, NULL, NULL);
```
Comment (oleavr) [inline gum/gum.c]: Missing blank line before this one.

Comment (oleavr) [inline gum/gum.c]: For consistency if would be good to name this `message_utf16`.

File: gum/gum.c [hunk 5]
```diff
@@ -506,6 +509,18 @@ gum_on_log_message (const gchar * log_domain,
...
+  if (utf16_message)
```
Comment (oleavr) [inline gum/gum.c]: No need to check the result, we are in control of the call-sites and can assume they give us valid UTF-8.

File: gum/gum.c (hunk 2 above)
Comment (erik-smit) [inline gum/gum.c]: It wouldn't make sense to have it output to console just in case people happen to be frida-ing a console-application?

I guess it makes sense to only have 1 definite path for those messages to come out.
Duplication is confusing.

Comment (erik-smit) [inline gum/gum.c]: Why sometimes the `#if defined` and sometimes the `# ifdef`?
And why sometimes the `#elif` and sometimes the `#else # ifdef`?

File: gum/gum.c [hunk 6]
```diff
@@ -16,6 +16,9 @@
 #include <ffi.h>
 #include <glib-object.h>
 #include <gio/gio.h>
+#ifdef HAVE_WINDOWS
+#  include <windows.h>
```
Comment (oleavr) [inline gum/gum.c]: Should only be one space per level here.

File: gum/gum.c [hunk 7]
```diff
@@ -387,7 +390,14 @@ gum_on_log_message (const gchar * log_domain,
                     const gchar * message,
                     gpointer user_data)
 {
-#ifdef HAVE_ANDROID
+#if defined(HAVE_WINDOWS)
+  gunichar2 * message_utf16;
```
Comment (oleavr) [inline gum/gum.c]: Missing blank line between declaration and first statement.

File: gum/gum.c [hunk 8]
```diff
@@ -387,7 +390,14 @@ gum_on_log_message (const gchar * log_domain,
...
+  message_utf16 = g_utf8_to_utf16 (message, -1, NULL, NULL, NULL);
+  OutputDebugString (message_utf16);
+  g_free (message_utf16);
+
+  return;
```
Comment (oleavr) [inline gum/gum.c]: This `return` and previous line should be omitted.

File: gum/gum.c (hunk 7 above)
Comment (erik-smit) [inline gum/gum.c]: fixed, but github doesn't 'see' it, because I've just added a blank line.

File: gum/gum.c [hunk 9]
```diff
@@ -506,9 +514,9 @@ gum_on_log_message (const gchar * log_domain,
 
     return;
   }
-  /* else: fall through to stdout/stderr logging */
-# endif
 
+  /* else: fall through to stdout/stderr logging */
+#else
```
Comment (oleavr) [inline gum/gum.c]: This should be reverted as it breaks the fallback on i/macOS.

File: gum/gum.c [hunk 10]
```diff
@@ -411,8 +420,7 @@ gum_on_log_message (const gchar * log_domain,
   }
 
   __android_log_write (priority, log_domain, message);
-#else
-# ifdef HAVE_DARWIN
+#elif defined(HAVE_DARWIN)
```
Comment (oleavr) [inline gum/gum.c]: This should be reverted as it breaks the fallback on i/macOS.

File: gum/gum.c [hunk 11]
```diff
@@ -387,7 +390,13 @@ gum_on_log_message (const gchar * log_domain,
                     const gchar * message,
                     gpointer user_data)
 {
-#ifdef HAVE_ANDROID
+#if defined(HAVE_WINDOWS)
```
Comment (oleavr) [inline gum/gum.c]: Missing space after `defined`.

File: gum/gum.c [hunk 12]
```diff
@@ -387,7 +390,13 @@ gum_on_log_message (const gchar * log_domain,
...
+  gunichar2 * message_utf16;
+
+  message_utf16 = g_utf8_to_utf16 (message, -1, NULL, NULL, NULL);
+  OutputDebugString (message_utf16);
+  g_free (message_utf16);
+#elif defined(HAVE_ANDROID)
```
Comment (oleavr) [inline gum/gum.c]: Missing space after `defined`.

File: gum/gum.c [hunk 13]
```diff
@@ -506,9 +515,9 @@ gum_on_log_message (const gchar * log_domain,
 
     return;
   }
+
   /* else: fall through to stdout/stderr logging */
 # endif
-
```
Comment (oleavr) [inline gum/gum.c]: This change is unrelated and should be reverted.

**Comment by oleavr** [comment]
Thanks! ❤️
